### PR TITLE
Expose units

### PR DESCRIPTION
**Pull request by MauriceHendrix**
## Description

I fixed --expose_units to include both derived quantities and modifiable parameters.
As I had to process the stimulus current in an awkward way, so I moved stimulus current conversion earlier (where we get the derivatives) and actually updated the model there instead of doing it just before printing.

The unit conversion logic is still just before printing, hopefully this can be retired later, when i start using cellmlmanip unit conversion.

I also changed the tests a little bit moving the logic of comparing a model against it's reference into chaste_test_utils

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have updated all documentation necessary.
- [ ] I have checked spelling in (new) comments.

## Testing
- [X] Testing is done automatically and codecov shows test coverage
- [ ] This cannot be tested automatically

### CODE REVIEW

File: weblab_cg/data/tests/chaste_reference_models/Normal/additional_models/Shannon2004.cpp [hunk 1]
```diff
@@ -871,7 +872,7 @@ void OdeSystemInformation<CellShannon2004FromCellML>::Initialise(void)
     this->mVariableUnits.push_back("millimolar");
     this->mInitialConditions.push_back(0.002143165);
     
-    this->mInitialised = true;
+    this->mInitialised = true;
```
Comment (jonc125) [inline weblab_cg/data/tests/chaste_reference_models/Normal/additional_models/Shannon2004.cpp]: Some extra whitespace creeping in here ;)

File: weblab_cg/_generator_chaste_cg.py [hunk 2]
```diff
@@ -91,11 +91,13 @@ def __init__(self, model, class_name, file_name):
         # Store default options
         self.class_name = class_name
         self.file_name = file_name
-        self.dynamically_loadable = False
-
         self.generated_hpp = ''
         self.generated_cpp = ''
+
         self._is_self_excitatory = False
+        self._dynamically_loadable = kwargs['dynamically_loadable'] if 'dynamically_loadable' in kwargs else False
```
Comment (jonc125) [inline weblab_cg/_generator_chaste_cg.py]: Better is `kwargs.get('dynamically_loadable', False)`

File: weblab_cg/_generator_chaste_cg.py [hunk 3]
```diff
@@ -205,9 +197,33 @@ def _add_units(self):
...
+                and q not in self._model.get_state_symbols()
+                and not q == self._time_variable]
+
     def _get_modifiable_parameters(self):
-        return self._model.get_symbols_by_rdf((self._PYCMLMETA,
-                                              'modifiable-parameter'), 'yes')
+        """ Get all modifiable parameters
+
+            Note: the result depends on self._expose_annotated_variables to determine whether or not to include
+            variables exposed with oxford metadata that are derived quantities but are not annotated as such"""
```
Comment (jonc125) [inline weblab_cg/_generator_chaste_cg.py]: s/derived quantities/parameters/

File: weblab_cg/_generator_chaste_cg.py [hunk 4]
```diff
@@ -205,9 +197,33 @@ def _add_units(self):
...
+        if self._expose_annotated_variables:
+            # Combined and sorted in document order
```
Comment (jonc125) [inline weblab_cg/_generator_chaste_cg.py]: Not doc order any more!

File: weblab_cg/_generator_chaste_cg.py [hunk 5]
```diff
@@ -400,9 +507,10 @@ def _get_derived_quant(self):
         if self._expose_annotated_variables:
             # Combined and sorted in document order
```
Comment (jonc125) [inline weblab_cg/_generator_chaste_cg.py]: Now name order

File: weblab_cg/_generator_chaste_cg.py [hunk 6]
```diff
@@ -580,107 +676,46 @@ def _get_desired_units(self, term):
     def _format_derivative_equations(self):
         """Format derivative equations for chaste output"""
         # exclude ionic currents
-        equations = []
-        d_eqs = [eq for eq in self._derivative_equations
-                 if eq not in self._default_stimulus.values()]
+#        d_eqs = [{'lhs': self._printer.doprint(eqs.lhs),
```
Comment (jonc125) [inline weblab_cg/_generator_chaste_cg.py]: Why is this here?

File: weblab_cg/templates/chaste/normal_model.cpp [hunk 7]
```diff
@@ -174,9 +174,13 @@ void OdeSystemInformation<{{class_name}}>::Initialise(void)
...
     this->mParameterUnits.push_back("{{param["units"]}}");
 
+    {% endfor %}{% for param in derived_quantities %}// Derived Quantity index [{{loop.index0}}]:
+    this->mParameterNames.push_back("{{param["name"]}}");
+    this->mParameterUnits.push_back("{{param["units"]}}");
+
     {% endfor %}{% for attr in named_attributes %}
     this->mAttributes["{{attr["name"]}}"] = {{attr["value"]}};
-    {% endfor %}this->mInitialised = true;
+    {% endfor %}this->mInitialised = true;
```
Comment (jonc125) [inline weblab_cg/templates/chaste/normal_model.cpp]: Here's the culprit!

File: weblab_cg/_generator_chaste_cg.py (hunk 6 above)
Comment (MauriceHendrix) [reply weblab_cg/_generator_chaste_cg.py]: Was getting ahead of myself and wrote how it would hopefully be after cellmlmanip unit conversion, but removed for now.
